### PR TITLE
Update user model,combined with profile

### PR DESCRIPTION
**Pull request by Swooty18**
**CHECK LIST**
- [ + ] specs added
- [ + ] specs passed
- [ + ] code coverage >90%
- [ + ] DB indexes added if needed
- [ + ] seed.rb file is updated if needed
- [ + ] PR is reviewed manually again (to make sure you have 100% ready code)
- [ + ] PR doesn't have conflicts with main branch
- [ + ] I've checked new feature as logged in and logged out user if needed
- [ + ] "let!" is used in specs
- [ - ] screenshot of feature is added if task has UI(hasn't UI)

### CODE REVIEW

File: app/models/osbb.rb [hunk 1]
```diff
@@ -5,6 +5,7 @@ class Osbb < ApplicationRecord
   validates :name, presence: true, length: { minimum: 3, maximum: 255 }
   validates :phone, format: { with: VALID_PHONE }
   validates :email, length: { maximum: 255 }, format: { with: VALID_EMAIL_REGEX }
+  has_many :users, dependent: :nullify
```
Comment (chu4) [inline app/models/osbb.rb:8]: change association name to members + add an association for lead member

File: app/models/user.rb [hunk 2]
```diff
@@ -2,8 +2,13 @@ class User < ApplicationRecord
   validates :first_name, presence: true, length: { maximum: 50 }
   validates :last_name, presence: true, length: { maximum: 50 }
   validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }, uniqueness: true, length: { maximum: 255 }
-
+  validates :mobile, presence: true,
+                     numericality: true,
+                     length: { minimum: 10, maximum: 20 }
   before_save :downcase_email
+  enum role: { admin: 0, lead: 1, members: 2, simple: 3 }
```
Comment (chu4) [inline app/models/user.rb:9]: make a constant list for roles

check how company types were made

File: db/migrate/20201006074043_add_fields_to_users.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class AddFieldsToUsers < ActiveRecord::Migration[6.0]
+  def change
+    add_column :users, :password, :string
+    add_column :users, :mobile, :string
+    add_column :users, :birthday, :date
+    add_column :users, :avatar, :string
```
Comment (chu4) [inline db/migrate/20201006074043_add_fields_to_users.rb:6]: not simply add a column, but a possibility add avatar as well)
use carrierwave or chose some other

synchronize with a teammate who implements User crud

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -58,6 +62,30 @@
...
+    end
+
+    context 'when mobile not valid' do
+      it 'reject mobile that has characters' do
+        user.mobile = '06866666ab'
+        expect(user).not_to be_valid
+      end
+
+      it 'reject mobile that has less 10 digits' do
+        user.mobile = '012345678'
```
Comment (BogdanBusko) [inline spec/models/user_spec.rb:84]: shoulda can do it to

File: app/models/user.rb [hunk 5]
```diff
@@ -2,8 +2,13 @@ class User < ApplicationRecord
   validates :first_name, presence: true, length: { maximum: 50 }
   validates :last_name, presence: true, length: { maximum: 50 }
   validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }, uniqueness: true, length: { maximum: 255 }
-
+  validates :mobile, presence: true,
```
Comment (IvanDanyliv) [inline app/models/user.rb:5]: add more validates

File: spec/models/user_spec.rb (hunk 4 above)
Comment (Swooty18) [reply spec/models/user_spec.rb:84]: What do you mean, maybe leave it as is?

File: db/migrate/20201006074043_add_fields_to_users.rb (hunk 3 above)
Comment (Swooty18) [reply db/migrate/20201006074043_add_fields_to_users.rb:6]: teammate who implements User crud said he would do it

File: app/models/user.rb (hunk 5 above)
Comment (Swooty18) [reply app/models/user.rb:21]: ok

File: app/models/user.rb (hunk 2 above)
Comment (Swooty18) [reply app/models/user.rb]: ok

File: app/models/osbb.rb (hunk 1 above)
Comment (Swooty18) [reply app/models/osbb.rb]: ok

File: spec/models/user_spec.rb (hunk 4 above)
Comment (BogdanBusko) [reply spec/models/user_spec.rb:90]: https://github.com/thoughtbot/shoulda-matchers use this gem for this validations

File: app/models/osbb.rb [hunk 6]
```diff
@@ -9,6 +9,8 @@ class Osbb < ApplicationRecord
   validates :email,  length: { maximum: 255 },
                      format: { with: VALID_EMAIL_REGEX,
                                message: 'format is not valid' }
+  has_many :members, class_name: :User, dependent: :nullify
+  has_one :lead, class_name: :User, dependent: :nullify
```
Comment (chu4) [inline app/models/osbb.rb:13]: make it this way `class_name: "User"`

association lead should be user with role LEAD

File: spec/models/user_spec.rb (hunk 4 above)
Comment (Swooty18) [reply spec/models/user_spec.rb]: refactor

File: app/models/osbb.rb (hunk 6 above)
Comment (Swooty18) [reply app/models/osbb.rb]: refactor

File: app/models/osbb.rb [hunk 7]
```diff
@@ -9,6 +9,9 @@ class Osbb < ApplicationRecord
   validates :email,  length: { maximum: 255 },
                      format: { with: VALID_EMAIL_REGEX,
                                message: 'format is not valid' }
+  has_many :members, class_name: "User", dependent: :nullify
+  has_one :lead, -> { where "role = #{User::ROLES.each_index.detect { |i| User::ROLES[i] == User::LEAD }}" },
```
Comment (chu4) [inline app/models/osbb.rb:13]: always use brackets where()

try not to interpolate in DB queries
(not really a case here but still as a reminder)
https://guides.rubyonrails.org/security.html#sql-injection 

anyway what is this??????
`User::ROLES.each_index.detect { |i| User::ROLES[i] == User::LEAD }`

you dont need a string query here use a hash way
https://apidock.com/rails/ActiveRecord/QueryMethods/where
`where(role: role_that_you_need)`

**Review by marikrebega — APPROVED** [review]

File: app/models/user.rb [hunk 8]
```diff
@@ -3,9 +3,30 @@ class User < ApplicationRecord
...
   validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }, uniqueness: true, length: { maximum: 255 }
+  validates :mobile, presence: true,
+                     numericality: true,
+                     length: { minimum: 10, maximum: 14 }
+  validates :sex, presence: true, length: { maximum: 10 }
+  validates :password, presence: true, length: { minimum: 5, maximum: 25 }
+
+  enum role: ROLES
+
+  belongs_to :osbb, optional: true
```
Comment (vasya539) [inline app/models/user.rb:29]: style guide. the association goes before  validations

File: spec/factories/users.rb [hunk 9]
```diff
@@ -1,9 +1,14 @@
 FactoryBot.define do
   factory :user do
+    email { Faker::Internet.email }
     first_name { Faker::Name.first_name }
     last_name  { Faker::Name.last_name }
-    email { Faker::Internet.email }
+    avatar { Faker::File.file_name }
```
Comment (vasya539) [inline spec/factories/users.rb:6]: no you would not get avatar that way. You need to create folder in spec directory where you store some random avatars, then you assign them here

Comment (Swooty18) [reply spec/factories/users.rb:6]: teammate who implements User crud said he would do it)

File: app/models/user.rb (hunk 8 above)
Comment (Swooty18) [reply app/models/user.rb:29]: ok

File: app/models/osbb.rb (hunk 7 above)
Comment (Swooty18) [reply app/models/osbb.rb]: refactor

**Review by DenisRem — APPROVED** [review]

**Review by chu4 — APPROVED** [review]

**Pull request merged**